### PR TITLE
Implementation of `app export` command

### PR DESCRIPTION
**Pull request by andreas-kupries**
Ref #1253 

- [x] Client side 
- [x] Server side
- [x] Tests
- [x] Swagger
- [x] values.yaml
- [x] chart tarball

### CODE REVIEW

File: internal/cli/usercmd/app.go [hunk 1]
```diff
@@ -201,6 +204,59 @@ func (c *EpinioClient) AppShow(appName string) error {
...
+	if err != nil {
+		return err
+	}
+
+	err = c.API.AppGetPart(c.Settings.Namespace, appName, "chart", filepath.Join(directory, "app-chart.tar.gz"))
+	if err != nil {
+		return err
+	}
+
+	//	err = c.API.AppGetPart(c.Settings.Namespace, appName, "image", filepath.Join(directory,"image"))
```
Comment (olblak) [inline internal/cli/usercmd/app.go:252]: Any reason for this block to be commented or is it just an error?

File: pkg/api/core/v1/client/apps.go [hunk 2]
```diff
@@ -138,6 +138,60 @@ func (c *Client) AppShow(namespace string, appName string) (models.App, error) {
...
+
+	defer response.Body.Close()
+	reqLog.V(1).Info("request finished")
+
+	// Create the file
+	out, err := os.Create(destinationPath)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
```
Comment (olblak) [inline pkg/api/core/v1/client/apps.go:185]: Don't you want to catch the potential error using something like 

```
defer func(){
 err = out.Close()
  if err != nil {
    fmt.Println(err)
  } 
}()
```

Comment (olblak) [reply pkg/api/core/v1/client/apps.go:185]: It's sometimes usefull to catch error from defer execution

File: internal/cli/usercmd/app.go (hunk 1 above)
Comment (andreas-kupries) [reply internal/cli/usercmd/app.go:252]: `image` is the one part I currently have no full idea how to handle server side. So I blocked it client side for now, while also not wanting to forget the possibility.

File: pkg/api/core/v1/client/apps.go (hunk 2 above)
Comment (andreas-kupries) [reply pkg/api/core/v1/client/apps.go:185]: Hm.

Comment (andreas-kupries) [reply pkg/api/core/v1/client/apps.go:192]: Not taking it.

File: internal/api/v1/application/part.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	if app == nil {
+		return apierror.AppIsNotKnown(appName)
+	}
+
+	if app.Workload == nil {
+		// While the app exists it has no workload, and therefore no chart to export
+		return apierror.NewBadRequest("No chart available for application without workload")
+	}
+
+	if partName == "chart" {
```
Comment (jimmykarily) [inline internal/api/v1/application/part.go:52]: Split in 3 function? E.g.

```
    switch partName {
    case "chart":
        fetchChart(ctx)
    case "image":
        fetchImage(ctx)
    case "values":
        fetchValues(ctx)
    }
```

File: internal/api/v1/application/part.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+		requestctx.Logger(c.Request.Context()).Info("OK",
+			"origin", c.Request.URL.String(),
+			"returning", fmt.Sprintf("%d bytes %s as is", contentLength, contentType),
+		)
+		c.DataFromReader(http.StatusOK, contentLength, contentType, reader, nil)
+		return nil
+	}
+
+	if partName == "image" {
+		return apierror.NewInternalError("image part not yet supported")
```
Comment (jimmykarily) [inline internal/api/v1/application/part.go:76]: Maybe an `HTTP 400` is more suitable here? `500`s are unexpected things usually.

File: internal/api/v1/application/part.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	}
+
+	if partName == "values" {
+		log := requestctx.Logger(ctx)
+
+		yaml, err := helm.Values(cluster, log, app.Meta)
+
```
Comment (jimmykarily) [inline internal/api/v1/application/part.go:83]:
```suggestion
```

File: internal/cli/usercmd/app.go [hunk 6]
```diff
@@ -201,6 +204,59 @@ func (c *EpinioClient) AppShow(appName string) error {
...
+	// 1. Create directory
+	// 2. Retrieve values.yaml for application
+	// 3. Retrieve chart tarball for application
+	// (4.) Retrieve application image (staged sources)
+	// Save all to the directory (in each step)
+	//
+	// While the `values.yaml` is likely quite small it is
+	// essentially a file like all the other parts (chart tarball,
+	// image). For consistency just handle all of them as files,
+	// in the same manner (web socket ?)
```
Comment (jimmykarily) [inline internal/cli/usercmd/app.go:235]:
```suggestion
```

The code is rather explanatory itself (like it should). I don't think we need to maintain comments as well.

File: internal/api/v1/application/part.go (hunk 5 above)
Comment (andreas-kupries) [reply internal/api/v1/application/part.go:87]: Done with commit 0f85444f1587cd6e32799c0ac2b12c093be2bd86

File: internal/cli/usercmd/app.go (hunk 6 above)
Comment (andreas-kupries) [reply internal/cli/usercmd/app.go]: Done with commit 0f85444f1587cd6e32799c0ac2b12c093be2bd86

File: internal/api/v1/application/part.go (hunk 4 above)
Comment (andreas-kupries) [reply internal/api/v1/application/part.go]: Done with commit 0f85444f1587cd6e32799c0ac2b12c093be2bd86

File: internal/api/v1/application/part.go (hunk 3 above)
Comment (andreas-kupries) [reply internal/api/v1/application/part.go]: Done with commit 0f85444f1587cd6e32799c0ac2b12c093be2bd86

**Review by jimmykarily — APPROVED** [review]

**Comment by gitricko** [comment]
@andreas-kupries : I tested the cli and saw the export. But i did not see oci images being exported. How do i use this command ?

**Comment by andreas-kupries** [comment]
The OCI image of the app is currently not exported. At the moment only the chart tarball and the configuration (values.yaml) are. IIRC @jimmykarily mentioned it in the demo that the image is not exported.

**Comment by jimmykarily** [comment]
@gitricko for now this feature is only useful if:
- The registry is external (thus the container images are accessible from anywhere)
- The image is exported to another registry manually + the helm chart is edited to point to that other registry

We decided it was ok for this the first iteration but we can improve it more by exporting the image too in the future (see my comment here: https://github.com/epinio/epinio/issues/1253#issuecomment-1066572188). Trying to think how users would use this feature, I feel it is a natural decision to keep the container registry accessible by both the epinio cluster and the other one (non-epinio) because it is simpler and doesn't have any serious drawbacks. If you think this assumption is wrong, please let us know and we will try to prioritize the exporting of the image.

**Comment by jimmykarily** [comment]
@gitricko feel free to add your comments here: https://github.com/epinio/epinio/issues/1352

**Pull request merged**